Added collection view component

aducret opened the pull request:
This PR add a collection view.

Example: 
``` swift code
collection(
    properties: properties(layoutValues: layoutValues) {
        let items: [CollectionItemProperties<Action>] = [
            collectionItem(onTap: .navigate(to: .lessonDetail), identifier: "cell") { _ in
                container(
                    children: [
                        label(text: "cell 1")
                    ],
                    style: styleSheet() { base in
                        base.backgroundColor = Color.red
                    }
                )
            }
        ]
        $0.items = items
        $0.showsVerticalScrollIndicator = false
        $0.showsHorizontalScrollIndicator = false
        $0.isSnapToCellEnabled = false
    },
    style: styleSheet() {
        $0.backgroundColor = .blue
    },
    layout: layout {
        $0.width = Dimension(value: 300)
        $0.height = Dimension(value: 150)
    }
)
```

# Screencast
![pr](https://cloud.githubusercontent.com/assets/1799612/24821150/4b0d8dd4-1bc3-11e7-99c5-cfab03eb3dfc.gif)
![pr](https://cloud.githubusercontent.com/assets/1799612/24821307/431477e0-1bc4-11e7-807a-09368ab82f46.gif)

Commit:
Added collection view component

guidomb commented on Sources/UIKit/PortalCollectionView.swift:
```diff
@@ -0,0 +1,162 @@
+    
+}
+
+public struct CollectionViewLayoutValues {
```
~~This strcut should be defined inside `Collection.swift` file and all dependencies with UIKit should be removed.~~ See my comment above about this type.

guidomb commented on Sources/Components/Collection.swift:
```diff
@@ -0,0 +1,93 @@
+public struct CollectionProperties<MessageType> {
+    
+    public var items: [CollectionItemProperties<MessageType>]
+    public var layoutValues: CollectionViewLayoutValues
```
I think it would be better not to have this property at all. For now we are not going to support any other type of collection layout so I would put all properties that are currently in `CollectionViewLayoutValues` directly here. Keep in mind that you have to remove any dependency with UIKit.

guidomb commented on Sources/UIKit/Renderers/CollectionRender.swift:
```diff
@@ -0,0 +1,31 @@
+    let layout: Layout
+    
+    func render(with layoutEngine: LayoutEngine, isDebugModeEnabled: Bool) -> Render<MessageType> {
+        let collectionView = PortalCollectionView(items: properties.items, layoutEngine: layoutEngine, layoutValues: properties.layoutValues)
```
After making the change I've suggested regarding the layout properties. Here I would create a CollectionViewFlowLayout based on those properties pass it as a parameter to `PortalCollectionView`

guidomb commented on Sources/UIKit/PortalCollectionView.swift:
```diff
@@ -0,0 +1,162 @@
+        self.dataSource = self
+        self.delegate = self
+        
+        let identifiers = Array(Set(items.map { $0.identifier }))
```
I don't think is necessary to convert the set to an array just to iterate over the elements.

guidomb commented on Sources/UIKit/PortalCollectionView.swift:
```diff
@@ -0,0 +1,162 @@
+fileprivate extension PortalCollectionView {
+    
+    fileprivate func scrollToItem(_ position: Int, animated: Bool) {
+        DispatchQueue.main.async(execute: {
```
Refactor this to

```swift
DispatchQueue.main.async {

}
```

guidomb replied:
Could you add a comment explaining why do you need to dispatch this block of code

guidomb commented on Sources/UIKit/PortalCollectionView.swift:
```diff
@@ -0,0 +1,162 @@
+
+    fileprivate func dequeueReusableCell(with identifier: String, for indexPath: IndexPath) -> PortalCollectionViewCell<MessageType> {
+        if let cell = dequeueReusableCell(withReuseIdentifier: identifier, for: indexPath) as? PortalCollectionViewCell<MessageType> {
+            cell.layoutEngine = layoutEngine
```
This does not seem to be necessary

aducret replied:
It is necessary, when we set this value it create a render

``` swift code
public var layoutEngine: LayoutEngine? = .none {
        didSet {
            if let layoutEngine = layoutEngine {
                self.renderer = UIKitComponentRenderer(containerView: contentView, layoutEngine: layoutEngine)
            }
        }
    }
```

guidomb replied:
I don't see the real need of that sideeffect. Once the renderer is created it can be re used.

Commit:
Fixs:

- Avoid ussing UIKit in Collection, now is used in CollectionRender.
- Little style refactor.

guidomb reviewed (commented):
@aducret The patch is almost ready to be merged. I've just made a comment regarding creating a new renderer each time. Can we address that please.

Commit:
Avoiding sideeffect in renderer creation

aducret commented:
@guidomb changes ammended last commit
